### ISSUE TITLE
Opengl linking errors on windows

### ISSUE THREAD

**Issue by ghost** [stackgl/headless-gl#25]
The addon is not built on windows out of the box. If you try to build it with vs2013 x64 like this:

> npm i gl --msvs_version=2013

You'll get a lot of opengl linker errors, what's the missing dependency? The readme states is just python2.7 and visual studio, both of them are satisfied

webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteTextures
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetIntegerv
webgl.obj : error LNK2001: unresolved external symbol __imp_glDepthFunc
webgl.obj : error LNK2001: unresolved external symbol __imp_glClear
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetBooleanv
webgl.obj : error LNK2001: unresolved external symbol __imp_glShaderSource
webgl.obj : error LNK2001: unresolved external symbol __imp_glClearColor
webgl.obj : error LNK2001: unresolved external symbol __imp_glCreateProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetShaderiv
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetShaderSource
webgl.obj : error LNK2001: unresolved external symbol __imp_glCheckFramebufferStatus
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetFramebufferAttachmentParameteriv
webgl.obj : error LNK2001: unresolved external symbol __imp_glDrawElements
webgl.obj : error LNK2001: unresolved external symbol __imp_glTexSubImage2D
webgl.obj : error LNK2001: unresolved external symbol __imp_eglDestroyContext
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glGenRenderbuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_glTexParameterf
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetVertexAttribiv
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniformMatrix4fv
webgl.obj : error LNK2001: unresolved external symbol __imp_glColorMask
webgl.obj : error LNK2001: unresolved external symbol __imp_glScissor
webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glVertexAttrib1f
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform1i
webgl.obj : error LNK2001: unresolved external symbol __imp_glGenBuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_glPixelStorei
webgl.obj : error LNK2001: unresolved external symbol __imp_glBlendFuncSeparate
webgl.obj : error LNK2001: unresolved external symbol __imp_glVertexAttrib3f
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetShaderInfoLog
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsEnabled
webgl.obj : error LNK2001: unresolved external symbol __imp_glDetachShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glUseProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniformMatrix3fv
webgl.obj : error LNK2001: unresolved external symbol __imp_glActiveTexture
webgl.obj : error LNK2001: unresolved external symbol __imp_glFinish
webgl.obj : error LNK2001: unresolved external symbol __imp_glBindBuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glBindTexture
webgl.obj : error LNK2001: unresolved external symbol __imp_glVertexAttrib2f
webgl.obj : error LNK2001: unresolved external symbol __imp_glPolygonOffset
webgl.obj : error LNK2001: unresolved external symbol __imp_eglMakeCurrent
webgl.obj : error LNK2001: unresolved external symbol __imp_glDisableVertexAttribArray
webgl.obj : error LNK2001: unresolved external symbol __imp_eglInitialize
webgl.obj : error LNK2001: unresolved external symbol __imp_eglChooseConfig
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetAttachedShaders
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform3f
webgl.obj : error LNK2001: unresolved external symbol __imp_glLinkProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glTexImage2D
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetError
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilOpSeparate
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform4i
webgl.obj : error LNK2001: unresolved external symbol __imp_glSampleCoverage
webgl.obj : error LNK2001: unresolved external symbol __imp_glDisable
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetRenderbufferParameteriv
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetString
webgl.obj : error LNK2001: unresolved external symbol __imp_glVertexAttribPointer
webgl.obj : error LNK2001: unresolved external symbol __imp_glDrawArrays
webgl.obj : error LNK2001: unresolved external symbol __imp_glCreateShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform2i
webgl.obj : error LNK2001: unresolved external symbol __imp_glDepthMask
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetVertexAttribfv
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetTexParameteriv
webgl.obj : error LNK2001: unresolved external symbol __imp_glCopyTexSubImage2D
webgl.obj : error LNK2001: unresolved external symbol __imp_glFramebufferTexture2D
webgl.obj : error LNK2001: unresolved external symbol __imp_glEnableVertexAttribArray
webgl.obj : error LNK2001: unresolved external symbol __imp_glTexParameteri
webgl.obj : error LNK2001: unresolved external symbol __imp_glVertexAttrib4f
webgl.obj : error LNK2001: unresolved external symbol __imp_glCopyTexImage2D
webgl.obj : error LNK2001: unresolved external symbol __imp_glClearStencil
webgl.obj : error LNK2001: unresolved external symbol __imp_glBindRenderbuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glReadPixels
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetFloatv
webgl.obj : error LNK2001: unresolved external symbol __imp_glGenFramebuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsRenderbuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glBindFramebuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilFunc
webgl.obj : error LNK2001: unresolved external symbol __imp_eglCreatePbufferSurface
webgl.obj : error LNK2001: unresolved external symbol __imp_glGenerateMipmap
webgl.obj : error LNK2001: unresolved external symbol __imp_glBlendFunc
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilMask
webgl.obj : error LNK2001: unresolved external symbol __imp_glCullFace
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetActiveUniform
webgl.obj : error LNK2001: unresolved external symbol __imp_glBufferSubData
webgl.obj : error LNK2001: unresolved external symbol __imp_glCompileShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilFuncSeparate
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform3i
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetAttribLocation
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform2f
webgl.obj : error LNK2001: unresolved external symbol __imp_glDepthRangef
webgl.obj : error LNK2001: unresolved external symbol __imp_glBlendColor
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsBuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetProgramInfoLog
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetVertexAttribPointerv
webgl.obj : error LNK2001: unresolved external symbol __imp_glViewport
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsTexture
webgl.obj : error LNK2001: unresolved external symbol __imp_eglCreateContext
webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glFramebufferRenderbuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glAttachShader
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsFramebuffer
webgl.obj : error LNK2001: unresolved external symbol __imp_glLineWidth
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetShaderPrecisionFormat
webgl.obj : error LNK2001: unresolved external symbol __imp_eglTerminate
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform4f
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetUniformfv
webgl.obj : error LNK2001: unresolved external symbol __imp_glIsProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glBufferData
webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteRenderbuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_glClearDepthf
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetUniformLocation
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetBufferParameteriv
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniform1f
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilOp
webgl.obj : error LNK2001: unresolved external symbol __imp_glBlendEquationSeparate
webgl.obj : error LNK2001: unresolved external symbol __imp_glEnable
webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteFramebuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_eglGetDisplay
webgl.obj : error LNK2001: unresolved external symbol __imp_glDeleteBuffers
webgl.obj : error LNK2001: unresolved external symbol __imp_glFrontFace
webgl.obj : error LNK2001: unresolved external symbol __imp_glFlush
webgl.obj : error LNK2001: unresolved external symbol __imp_glStencilMaskSeparate
webgl.obj : error LNK2001: unresolved external symbol __imp_glGenTextures
webgl.obj : error LNK2001: unresolved external symbol __imp_glRenderbufferStorage
webgl.obj : error LNK2001: unresolved external symbol __imp_glBlendEquation
webgl.obj : error LNK2001: unresolved external symbol __imp_glValidateProgram
webgl.obj : error LNK2001: unresolved external symbol __imp_glHint
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetActiveAttrib
webgl.obj : error LNK2001: unresolved external symbol __imp_glGetProgramiv
webgl.obj : error LNK2001: unresolved external symbol __imp_glUniformMatrix2fv
webgl.obj : error LNK2001: unresolved external symbol __imp_glBindAttribLocation

**Comment by mikolalysenko** [comment]
I want to investigate this, but I don't have access to a windows environment.  It could be that one of the .gyp files is broken on windows.  Right now I am trying to get a parallels vm with windows 10 working, but the visual studios set up keeps crashing.

**Comment by timknip** [comment]
Hi Mikola,

I got a pretty good windows environment. But stuck also at above linker errors.

Now: ANGLE compiles with no problem.

My main question is: what lib is the project trying to link to? The lib obviously isn't found (or there's some x64 / win32 mix up)

On thought I had: ANGLE builds to Release_x64, which is kind of a non-standard output directory afaik.

Anyway: if you have any pointers, let me know.

.

**Comment by mikolalysenko** [comment]
I think that you will need to modify the gypfiles to get this to work.  Take a look in:

https://github.com/stackgl/headless-gl/blob/master/binding.gyp
https://github.com/mikolalysenko/angle/tree/master/build
https://github.com/mikolalysenko/angle/blob/master/src/compiler.gypi
https://github.com/mikolalysenko/angle/blob/master/src/angle.gyp

The problem is that it is trying to link angle as a dynamic library, when we really want to use it as a static library.  There are some inconsistencies between the version of gyp used by node-gyp and the one that angle/chromium use to build their project.  Fixing this issue will probably require patching this up.  I think that some of the windows specific gyp stuff in the core binding.gyp is out of date and needs to be fixed.

**Comment by timknip** [comment]
Ok, will take a look coming days.

**Comment by timknip** [comment]
@mikolalysenko I was wondering on the 'deps' directory referenced in binding.gyp ... any other deps apart from glew?

**Comment by bpedrosa** [comment]
@mikolalysenko Not sure if related but the logs will also show:

npm ERR! Tell the author that this fails on your system:
npm ERR!     prebuild --download

Why is the prebuild command failing?

**Comment by mikolalysenko** [comment]
@timknip that deps stuff is legacy from the old node-webgl build scripts.  I haven't touched it since switching to ANGLE, but it probably doesn't need to be there.

**Comment by mikolalysenko** [comment]
@bpedrosa ignore the prebuild issue, that is because I haven't gotten windows to build in the first place!  Once that much works, then everything should be fine.

**Comment by timknip** [comment]
@mikolalysenko I followed https://chromium.googlesource.com/angle/angle/+/master/doc/DevSetup.md and build Release_x64. After that I edited headless-gl's binding.gyp to refer to that build. ```node example/fbo``` actually worked :-) However: only when copying libEGL.dll and libGLESv2.dll to /build/Release...

```['OS=="win"',
          {
            'include_dirs': [
              'd:/floorplanner/tim/angle/include'
              ],
            'library_dirs': [
              '../angle/build/Release_x64/lib'
              ],
            'libraries': [
              'libEGL.lib',
              'libGLESv2.lib'
              ],
            'defines' : [
              'WIN32_LEAN_AND_MEAN',
              'VC_EXTRALEAN',
              'GL_GLEXT_PROTOTYPES'
            ],
            'cflags' : [
              '/O2','/Oy','/GL','/GF','/Gm-','/EHsc','/MT','/GS','/Gy','/GR-','/Gd','/wd"4530"','/wd"4251"'
            ],
            'ldflags' : [
              '/OPT:REF','/OPT:ICF','/LTCG'
            ]
          }
        ]```

Now... of course... this doesn't solve anything, it just shows it works with the DLLs present...
it looks like the LIBs need the DLL to function (ie: the static lib needs the dll)

This is really above my head, No clue how to create the libs without needing the DLLs.

Guess the magic maybe is in ```BUILD.gn```, but that's just a hunch.

So... basically I'm throwing the towel,,,  Sorry.

**Comment by mikolalysenko** [comment]
Fixed in 4.0.0
